sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -1258,4 +1263,57 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+
+  test("SPARK-32932: Do not use local shuffle reader at final stage on write command") {
+    withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> PartitionOverwriteMode.DYNAMIC.toString,
+      SQLConf.SHUFFLE_PARTITIONS.key -> "5",
+      SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {
+      val data = for (
+        i <- 1L to 10L;
+        j <- 1L to 3L
+      ) yield (i, j)
+
+      val df = data.toDF("i", "j")
+        .repartition($"j")
+
+      withTable("t") {
+        df.write
+          .partitionBy("j")
+          .saveAsTable("t")
+        assert(spark.read.table("t").inputFiles.length == 3)
```

Comment:
it's a bit tricky to check the number of files. 3 distinct values don't always mean 3 files.

Can we use `QueryExecutionListener` to catch the logical plan, and check there is no local shuffle reader?

Reply:
Tests have been updated with `QueryExecutionListener` to check local shuffle reader. Please help review again.